### ISSUE TITLE
Test all architecture specific optimizations within the testsuite.

### ISSUE THREAD

**Issue by sebastianas** [openssl/openssl#23057]
x86-64 among other architectures has optimizations which are always compiled but only used during runtime based on CPU support such as SSE2 or AVX512. This leads to the problem that the testsuite only tests the "highest" available optimization on a certain CPU and never all of them including the not optimized "default" version. This means the building-box of distro tests the AVX512 version of aes-gcm while most of the users don't have AVX512 and use the untested version.

It would be nice if the testsuite could print optimizations which are about to be used, disable one and run the testsuite again.
This would cover all available CPU-optimization including pure software support.

**Comment by nhorman** [comment]
I belive all the test suite that actually run openssl apps now have a "get cpu" step that prints the processor flags of the cpu the test suite runs on.  Is that sufficient, or are you looking for a report that will translate that into what accelerations will be used for a given build?

**Comment by tom-cosgrove-arm** [comment]
There's a further potential issue here - some optimisations are only used for longer data lengths, so again, are not always run

**Comment by t8m** [comment]
I do not think it is feasible to run the full testsuite that many times per each change. On the other hand it would be beneficial to at least run the evp_test clearing various bits of CPU support - we would have to identify which bits are particularly useful to be cleared - there are too many to do one by one.

**Comment by t8m** [comment]
> There's a further potential issue here - some optimisations are only used for longer data lengths, so again, are not always run

We should identify if there are some longer data tests missing for algorithms that have special code-paths for these longer data and add such testcases to evp_test.

**Comment by nhorman** [comment]
what if we augment a ci job with appropriate gcc --coverage options, and used the output data from the test run to match against known symbols that are built against symbols we know to be compiled from accelerated source variants that are hardware accelerated?  i.e. if profile output data shows we executed the function aesni_cbc_encrypt, and the configuration shows that aesni_cbc_encrypt was built from the aensi_x86_64.S generated source, then we can state aesni_cbc_encrypt was tested for at least some data.  If we diff the set of executed functions against a full list of known accelerations for the arch, we would know where our test gaps are, without having to do unneeded test runs, or builds.

**Comment by kroeckx** [comment]
I think we should have a list of all the different implementations we have, including C and assembler versions. Try to test all of them, and print which have been tested successfully, which we didn't test because the CPU doesn't support it, and which we didn't test because it's disabled at compile time.

We should try to cover all of them in CI. For some things we have multiple C versions, but the compiled version only supports one of them.

I think it's also a good idea to actually have them documented.

**Comment by t8m** [comment]
@nhorman the problem is that the assembler implementations also have multiple codepaths and these are not covered by gcc --coverage data. Finding out what are the actual codepath variants based on CPU versions in the assembler sources will be the most demanding task.

**Comment by davidben** [comment]
See also https://github.com/openssl/openssl/discussions/22056 which discusses using Intel SDE.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
